x-pack/plugins/security_solution/public/common/components/exceptions/helpers.tsx

```diff
@@ -192,3 +222,121 @@ export const getFormattedComments = (comments: CommentsArray): EuiCommentProps[]
     timelineIcon: <EuiAvatar size="l" name={comment.created_by.toUpperCase()} />,
     children: <EuiText size="s">{comment.comment}</EuiText>,
   }));
+
+export const getFormattedBuilderEntries = (
+  indexPattern: IIndexPattern,
+  entries: BuilderEntry[]
+): FormattedBuilderEntry[] => {
+  const { fields } = indexPattern;
+  return entries.map((item) => {
+    if (entriesNested.is(item)) {
+      return {
+        parent: item.field,
+        operator: isOperator,
+        nested: getFormattedBuilderEntries(indexPattern, item.entries),
+        field: undefined,
+        value: undefined,
+      };
+    } else {
+      const [selectedField] = fields.filter(
+        ({ name }) => item.field != null && item.field === name
+      );
+      return {
+        field: selectedField,
+        operator: getExceptionOperatorSelect(item),
+        value: getEntryValue(item),
+      };
+    }
+  });
+};
+
+export const getValueFromOperator = (
+  field: IFieldType | undefined,
+  selectedOperator: OperatorOption
+): Entry => {
+  const fieldValue = field != null ? field.name : '';
+  switch (selectedOperator.type) {
+    case 'match':
+      return {
+        field: fieldValue,
+        type: OperatorTypeEnum.MATCH,
+        operator: selectedOperator.operator,
+        value: '',
+      };
+    case 'match_any':
+      return {
+        field: fieldValue,
+        type: OperatorTypeEnum.MATCH_ANY,
+        operator: selectedOperator.operator,
+        value: [],
+      };
+    case 'list':
+      return {
+        field: fieldValue,
+        type: OperatorTypeEnum.LIST,
+        operator: selectedOperator.operator,
+        list: { id: '', type: 'ip' },
+      };
+    default:
+      return {
+        field: fieldValue,
+        type: OperatorTypeEnum.EXISTS,
+        operator: selectedOperator.operator,
+      };
+  }
+};
+
+export const getNewExceptionItem = ({
+  listType,
+  listId,
+  namespaceType,
+  ruleName,
+}: {
+  listType: 'detection' | 'endpoint';
+  listId: string;
+  namespaceType: NamespaceType;
+  ruleName: string;
+}): CreateExceptionListItemBuilderSchema => {
+  return {
+    _tags: [listType],
+    comments: [],
+    description: `${ruleName} - exception list item`,
+    entries: [
+      {
+        field: '',
+        operator: 'included',
+        type: 'match',
+        value: '',
+      },
+    ],
+    item_id: undefined,
+    list_id: listId,
+    meta: {
+      temporaryUuid: uuid.v4(),
+    },
+    name: `${ruleName} - exception list item`,
+    namespace_type: namespaceType,
+    tags: [],
+    type: 'simple',
+  };
+};
+
+export const filterExceptionItems = (
```

Comment:
This is to ensure that we only bubble up valid exception item and entries. A user could add 10 blank rows or entries, this strips those out.